File: ios/Classes/handler/SearchDevicesHandlerImpl.swift
//
//  SearchDevicesHandlerImpl.swift
//  lan_scan
//
//  Created by arthur on 2024/4/7.
//

import Foundation
import Flutter

class SearchDevicesHandlerImpl: NSObject, FlutterStreamHandler {
    
    let wireless =  Wireless()
    
    func onListen(withArguments arguments: Any?, eventSink: @escaping FlutterEventSink) -> FlutterError? {
      
        guard wireless.isWiFiConnected()else{
            let notConnectedError = HostsScanError.notConnected
            eventSink(FlutterError(code: notConnectedError.rawValue, message: notConnectedError.description, details: nil))
            eventSink(FlutterEndOfEventStream)
            return nil
        }
        
        guard let ipv4 = wireless.getInternalWifiIpAddress(),
              let cidrPrefixLength = wireless.getInternalWifiCidrPrefixLength() else{
            let notConnectedError = HostsScanError.notConnected
            eventSink(FlutterError(code: notConnectedError.rawValue, message: notConnectedError.description, details: nil))
            eventSink(FlutterEndOfEventStream)
            return nil
        }
        
        let scanHostsAsyncTask = ScanHostsAsyncTask(eventSink: eventSink)
        Task{
            await scanHostsAsyncTask.scanHosts(ipv4: ipv4, cidrPrefixLength: cidrPrefixLength, timeout: 5000)
        }
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        return nil
    }
}

File: android/src/main/kotlin/com/example/lan_scan/handler/SearchDevicesHandlerImpl.kt
package com.example.lan_scan.handler

import android.content.Context
import android.text.format.DateUtils
import com.example.lan_scan.HostScanError
import com.example.lan_scan.network.Wireless
import com.example.lan_scan.task.ScanHostsAsyncTask
import io.flutter.plugin.common.EventChannel

class SearchDevicesHandlerImpl(context: Context) : EventChannel.StreamHandler {

    private val wireless = Wireless(context)

    override fun onListen(arguments: Any?, events: EventChannel.EventSink) {

        if (!wireless.isEnabled()) {
            events.error(
                HostScanError.WIFI_NOT_AVAILABLE.code,
                HostScanError.WIFI_NOT_AVAILABLE.description,
                null,
            )
            events.endOfStream()
            return
        }

        if (!wireless.isConnectedWifi()) {
            events.error(
                HostScanError.WIFI_NOT_CONNECTED.code,
                HostScanError.WIFI_NOT_CONNECTED.description,
                null,
            )
            events.endOfStream()
            return
        }


        val localIp = wireless.getInternalWifiIpAddress() // 获取内网IP
        val cidrPrefixLength = wireless.getInternalWifiCidrPrefixLength() // CIDR 前缀长度

        val scanHostsAsyncTask = ScanHostsAsyncTask(events)
        scanHostsAsyncTask.scanHosts(localIp, cidrPrefixLength, 5*DateUtils.MINUTE_IN_MILLIS)

    }

    override fun onCancel(arguments: Any?) {
    }
}